src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Value.WriteTests.cs

```diff
@@ -187,6 +187,46 @@ public static void TimeOnly_Write_Success(string value, string? expectedValue =
             string json = JsonSerializer.Serialize(ts);
             Assert.Equal($"\"{expectedValue ?? value}\"", json);
         }
+
+        [Fact]
+        public static void Int128_AsDictionaryKey_SerializesCorrectly()
+        {
+            // Regression test for https://github.com/dotnet/runtime/issues/116855
+            // Test with a single key to ensure deterministic output for Assert.Equal
+            var dict = new Dictionary<Int128, string>
+            {
+                [42] = "Answer"
+            };
+
+            string json = JsonSerializer.Serialize(dict);
+            
+            // Verify the exact JSON output - should be clean without junk data
+            Assert.Equal("{\"42\":\"Answer\"}", json);
+            
+            // E2E validation: should roundtrip correctly
+            var deserialized = JsonSerializer.Deserialize<Dictionary<Int128, string>>(json);
+            Assert.Equal(dict, deserialized);
+        }
+
+        [Fact]
+        public static void UInt128_AsDictionaryKey_SerializesCorrectly()
+        {
+            // Regression test for https://github.com/dotnet/runtime/issues/116855
+            // Test with a single key to ensure deterministic output for Assert.Equal
```

Comment:
For the case of dictionary, output is deterministic even with multiple keys. Please reinstate the keys as they were defined earlier.